src/disco/FixedWidthField/Real/ENDF/src/write.hpp

```diff
@@ -10,22 +10,88 @@ write( Representation real, Iterator& it ){
   } else if ( isInvalid( real ) ){
     writeInvalid( it );
   } else {
-    std::array< char, w + 3 > buffer;
-    auto greedyExponent = requiresTwoDigitExponent( real );
-    const auto length = w + 2;
-    const auto decimalDigits = 6 - greedyExponent;
-    int consumedCharacters = snprintf( &buffer[0], length + 1, "%*.*E",
-                                       length, decimalDigits, real );
-    if ( consumedCharacters < 0 ){
-      throw std::exception();
+
+    // characters excluded for fixed notation precision (initial space, digit,
+    // decimal point)
+    const unsigned int NUMBER_EXCLUDED_CHARS = 3;
+
+    // roundoff limit
+    const double ROUNDOFF_LIMIT = 0.5;
+
+    // Minimal exponent size is 2 (one sign and one digit)
+    constexpr int MIN_EXPWIDTH = 2;
+    // Minimal exponent for which we can use fixed notation
+    constexpr int MIN_EXPONENT = -2;
+
+    // Decompose real
+    double significand{ real };
+    int exponent{ 0 };
+    int expWidth{ MIN_EXPWIDTH };
+    if( real != 0.0 ){
+      // log10(significand 10^exponent) = exponent + log10(significand) and
+      // log10(significand) is within [0,1] so that exponent is given by the
+      // floor value
+      exponent = static_cast< int >( 
+        std::floor( std::log10( std::abs( real ) ) ) );
+
+      if( 0 != exponent ){
+        significand /= std::pow( 10.0, exponent );
+        expWidth += static_cast< int >( 
+            std::floor( std::log10( std::abs( exponent ) ) ) );
+      }
+      // if( 0 < exponent ){
+      //   significand /= std::pow( 10.0, exponent );
+      //   expWidth += static_cast< int >( 
+      //       std::floor( std::log10( std::abs( exponent ) ) ) );
+      // } else if( 0 > exponent ){
+      //   significand *= std::pow( 10.0, -exponent );
+      //   expWidth += static_cast< int >( 
+      //       std::floor( std::log10( std::abs( exponent ) ) ) );
+      // }
+    }
+
+    unsigned int width = w - expWidth;
+    unsigned int precision = width - NUMBER_EXCLUDED_CHARS;
+    bool fixed = false;
+
+    if( ( MIN_EXPONENT < exponent ) and ( exponent < ( w - 1 ) ) ){
+      double max = std::pow( 10.0, expWidth );
+      double remainder = std::abs( significand ) * std::pow( 10.0, precision );
+      remainder -= std::floor( remainder );
+      remainder *= max;
+
+      if( ( ROUNDOFF_LIMIT <= remainder ) and 
```

Comment:
You and I don't agree on how to round a 5, but I admit there isn't a true "correct" way to do it.  I was taught way back in grade school to round 5s up, you must have been taught down.  People that do a lot of rounding sometimes do crazy things like flipping a coin or rounding to the nearest even number for 5s.

Ultimately, my complaint here is I don't see a test with a 5.  And if we're going to do something potentially controversial, we should be pedantic enough about it to include an explicit test.

Reply:
Actually, I think you round up your 5 here on second look.  I just got confused by the order of the comparison.  But my point stands, include a test! :smile:

Reply:
I did what @whaeck did in his implementation. I just added a couple more tests.

Reply:
So for the record: this is for deciding whether or not to use fixed notation. The rounding itself (either in the fixed or scientific notation) is done when streaming the true value.

Reply:
If I remember correctly, the test using the remainder and the limit of 0.5 is to make sure that the fixed precision actually adds meaningful significant numbers.

Reply:
From the additional test implemented by @jlconlin, up to and including 0.5 is rounded down in the streaming operator.

Reply:
I did a test separately to see what's going on.  Letting C++ (or perhaps the compiler?) do the rounding actually uses fancy rounding rules.  `x.5` is rounded down if `x` is even and rounded up if `x` is odd.  `x.51` is rounded up.

```
   2.055 -->     2.06
   2.054 -->     2.05
  2.0551 -->     2.06
   2.065 -->     2.06
   2.064 -->     2.06
  2.0651 -->     2.07
```

I certainly wouldn't mess with this behavior, as this is a pretty justifiable algorithm!  And even if not all compilers were consistent, I still wouldn't mess with it.